File: tests/fastapi-framework/fastapi-framework/app/core/monitoring.py
"""
日誌與監控系統
"""
import time
import logging
import json
from datetime import datetime
from pathlib import Path
from collections import defaultdict
from typing import Dict
import asyncio
import threading

# 配置日誌
LOG_DIR = Path("logs")
LOG_DIR.mkdir(exist_ok=True)

logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s',
    handlers=[
        logging.FileHandler(LOG_DIR / "app.log"),
        logging.StreamHandler()
    ]
)
from app.core.config import get_settings as _get_settings
logger = logging.getLogger(_get_settings().APP_NAME)

# 內存指標存儲
class Metrics:
    """簡單的內存指標收集器"""
    _lock = threading.Lock()
    _data = {
        "requests": defaultdict(int),      # 各端點請求次數
        "errors": defaultdict(int),        # 各端點錯誤次數
        "durations": defaultdict(list),   # 各端點響應時間
        "ws_connections": 0,               # WebSocket 連接數
        "ws_messages": 0,                  # WebSocket 消息數
    }
    
    @classmethod
    def record_request(cls, endpoint: str, duration: float, status: int):
        with cls._lock:
            cls._data["requests"][endpoint] += 1
            if status >= 400:
                cls._data["errors"][endpoint] += 1
            cls._data["durations"][endpoint].append(duration)
            # 只保留最近1000條
            if len(cls._data["durations"][endpoint]) > 1000:
                cls._data["durations"][endpoint] = cls._data["durations"][endpoint][-1000:]
    
    @classmethod
    def record_ws_connect(cls):
        with cls._lock:
            cls._data["ws_connections"] += 1
    
    @classmethod
    def record_ws_message(cls):
        with cls._lock:
            cls._data["ws_messages"] += 1
    
    @classmethod
    def get_stats(cls) -> dict:
        with cls._lock:
            stats = {}
            for endpoint, count in cls._data["requests"].items():
                durations = cls._data["durations"].get(endpoint, [])
                errors = cls._data["errors"].get(endpoint, 0)
                
                avg_duration = sum(durations) / len(durations) if durations else 0
                p95_duration = sorted(durations)[int(len(durations) * 0.95)] if durations else 0
                
                stats[endpoint] = {
                    "requests": count,
                    "errors": errors,
                    "error_rate": round(errors / count * 100, 2) if count > 0 else 0,
                    "avg_duration_ms": round(avg_duration * 1000, 2),
                    "p95_duration_ms": round(p95_duration * 1000, 2),
                }
            
            return {
                "endpoints": stats,
                "websocket": {
                    "total_connections": cls._data["ws_connections"],
                    "total_messages": cls._data["ws_messages"],
                },
                "timestamp": datetime.now().isoformat()
            }
    
    @classmethod
    def reset(cls):
        with cls._lock:
            cls._data = {
                "requests": defaultdict(int),
                "errors": defaultdict(int),
                "durations": defaultdict(list),
                "ws_connections": 0,
                "ws_messages": 0,
            }


def log_request(endpoint: str, duration: float, status: int, method: str = "GET"):
    """記錄請求日誌"""
    # 記錄到 Metrics
    Metrics.record_request(endpoint, duration, status)
    
    # 結構化日誌
    log_data = {
        "timestamp": datetime.now().isoformat(),
        "method": method,
        "endpoint": endpoint,
        "duration_ms": round(duration * 1000, 2),
        "status": status,
    }
    
    if status >= 500:
        logger.error("Server Error: %s", json.dumps(log_data))
    elif status >= 400:
        logger.warning("Client Error: %s", json.dumps(log_data))
    else:
        logger.info("%s %s - %s - %.1fms", method, endpoint, status, duration * 1000)


def log_event(event_type: str, data: dict = None):
    """記錄業務事件"""
    log_data = {
        "timestamp": datetime.now().isoformat(),
        "event": event_type,
        "data": data or {}
    }
    logger.info(f"Event: {json.dumps(log_data)}")

File: server/core/monitoring.ts
/**
 * server/core/monitoring.ts — Layer 6 監控觀測層
 *
 * 對應 SmartClean: app/core/monitoring.py
 * 修復 Bug #10：_data 改為實例變量（非類變量），避免跨實例共享
 *
 * 功能：
 *   - Metrics：各端點請求數、錯誤數、p95 延遲
 *   - logger：結構化 JSON 日誌，同時輸出 console + logs/app.log
 *   - requestLogger：Express 中間件，自動記錄每個請求
 *   - logEvent：業務事件日誌
 */

import fs from 'fs';
import path from 'path';
import type { Request, Response, NextFunction } from 'express';

// ─── 日誌目錄 ──────────────────────────────────────────────────────────────
const LOG_DIR = 'logs';
if (!fs.existsSync(LOG_DIR)) fs.mkdirSync(LOG_DIR, { recursive: true });

const logStream = fs.createWriteStream(path.join(LOG_DIR, 'app.log'), { flags: 'a' });

type LogLevel = 'info' | 'warn' | 'error' | 'debug';

function write(level: LogLevel, message: string, data?: unknown): void {
  const line = JSON.stringify({
    timestamp: new Date().toISOString(),
    level,
    message,
    ...(data !== undefined ? { data } : {}),
  });
  logStream.write(line + '\n');
  if (level === 'error') console.error(line);
  else if (level === 'warn') console.warn(line);
  else if (level === 'debug') {
    if (process.env.NODE_ENV !== 'production') console.debug(line);
  } else console.log(line);
}

/** 結構化 JSON 日誌（對應 SmartClean logger = logging.getLogger(...)） */
export const logger = {
  info:  (message: string, data?: unknown) => write('info',  message, data),
  warn:  (message: string, data?: unknown) => write('warn',  message, data),
  error: (message: string, data?: unknown) => write('error', message, data),
  debug: (message: string, data?: unknown) => write('debug', message, data),
};

// ─── Metrics 收集器 ────────────────────────────────────────────────────────

interface EndpointStat {
  requests:  number;
  errors:    number;
  durations: number[]; // ms
}

/** 內存指標收集器（單進程）
 *  Fix Bug #10：使用 Map 實例變量，非類靜態變量
 */
class Metrics {
  private readonly data = new Map<string, EndpointStat>();
  private readonly start = Date.now();

  recordRequest(endpoint: string, durationMs: number, status: number): void {
    let stat = this.data.get(endpoint);
    if (!stat) {
      stat = { requests: 0, errors: 0, durations: [] };
      this.data.set(endpoint, stat);
    }
    stat.requests++;
    if (status >= 400) stat.errors++;
    stat.durations.push(durationMs);
    // 只保留最近 1000 條，防止無限增長
    if (stat.durations.length > 1000) stat.durations.splice(0, stat.durations.length - 1000);
  }

  getStats(): {
    uptime_seconds: number;
    endpoints: Record<string, unknown>;
    timestamp: string;
  } {
    const endpoints: Record<string, unknown> = {};
    for (const [ep, stat] of this.data) {
      const sorted = [...stat.durations].sort((a, b) => a - b);
      const avg = sorted.length ? sorted.reduce((s, v) => s + v, 0) / sorted.length : 0;
      const p95 = sorted.length ? sorted[Math.floor(sorted.length * 0.95)] : 0;
      endpoints[ep] = {
        requests:         stat.requests,
        errors:           stat.errors,
        error_rate:       stat.requests > 0 ? +((stat.errors / stat.requests) * 100).toFixed(2) : 0,
        avg_duration_ms:  +avg.toFixed(2),
        p95_duration_ms:  +p95.toFixed(2),
      };
    }
    return {
      uptime_seconds: Math.floor((Date.now() - this.start) / 1000),
      endpoints,
      timestamp: new Date().toISOString(),
    };
  }

  /** Prometheus 文本格式（對應 SmartClean /api/monitoring/metrics） */
  prometheusFormat(appName = 'stock_signal'): string {
    const name  = appName.toLowerCase().replace(/\s+/g, '_');
    const stats = this.getStats();
    const lines: string[] = [
      `# HELP ${name}_http_requests_total Total HTTP requests`,
      `# TYPE ${name}_http_requests_total counter`,
    ];
    for (const [ep, d] of Object.entries(stats.endpoints as Record<string, any>)) {
      lines.push(`${name}_http_requests_total{endpoint="${ep}"} ${d.requests}`);
    }
    lines.push(
      '',
      `# HELP ${name}_uptime_seconds Process uptime`,
      `# TYPE ${name}_uptime_seconds gauge`,
      `${name}_uptime_seconds ${stats.uptime_seconds}`,
    );
    return lines.join('\n');
  }

  reset(): void {
    this.data.clear();
  }
}

/** 全局單例 */
export const metrics = new Metrics();

// ─── Express 請求日誌中間件 ────────────────────────────────────────────────

/**
 * requestLogger — 自動記錄每個 HTTP 請求
 * 掛在 express.json() 之後，路由之前
 */
export function requestLogger(req: Request, res: Response, next: NextFunction): void {
  const startMs = Date.now();
  res.on('finish', () => {
    const durationMs = Date.now() - startMs;
    // 優先用 route.path（有 params 時更易讀），fallback 到 req.path
    const endpoint = `${req.method} ${req.route?.path ?? req.path}`;
    metrics.recordRequest(endpoint, durationMs, res.statusCode);

    if (res.statusCode >= 500) {
      logger.error(`${endpoint} ${res.statusCode}`, { durationMs, ip: req.ip });
    } else if (res.statusCode >= 400) {
      logger.warn(`${endpoint} ${res.statusCode}`, { durationMs, ip: req.ip });
    } else {
      logger.info(`${endpoint} ${res.statusCode} ${durationMs}ms`);
    }
  });
  next();
}

// ─── 業務事件日誌 ──────────────────────────────────────────────────────────

/** 記錄業務事件（對應 SmartClean log_event）*/
export function logEvent(eventType: string, data?: Record<string, unknown>): void {
  logger.info(`[Event] ${eventType}`, data);
}
